store/store_download.go

```diff
@@ -253,6 +259,109 @@ func downloadReqOpts(storeURL *url.URL, cdnHeader string, opts *DownloadOptions)
 	return &reqOptions
 }
 
+type downloadTimeoutError struct {
+	// for testing, to inspect speed
+	Speed float64
+}
+
+func (e *downloadTimeoutError) Error() string {
+	return fmt.Sprintf("download timed out")
+}
+
+// implements io.Writer interface
+// XXX: move to osutil?
+type TransferSpeedMonitoringWriter struct {
+	mu sync.Mutex
+
+	measureTimeWindow   time.Duration
+	minDownloadSpeedBps float64
+
+	ctx context.Context
+
+	// internal state
+	start   time.Time
+	written int
+	cancel  func()
+	err     error
+
+	// for testing
+	measuredWindows int
+}
+
+// NewTransferSpeedMonitoringWriterAndContext returns an io.Writer that measures write speed
+// in measureTimeWindow windows an cancels the operation if minDownloadSeepdBps
+// is not achieved.
+// Monitor() must be called to start actual measurement.
+func NewTransferSpeedMonitoringWriterAndContext(origCtx context.Context, measureTimeWindow time.Duration, minDownloadSpeedBps float64) (*TransferSpeedMonitoringWriter, context.Context) {
+	ctx, cancel := context.WithCancel(origCtx)
+	w := &TransferSpeedMonitoringWriter{
+		measureTimeWindow:   measureTimeWindow,
+		minDownloadSpeedBps: minDownloadSpeedBps,
+		ctx:                 ctx,
+		cancel:              cancel,
+	}
+	return w, ctx
+}
+
+func (w *TransferSpeedMonitoringWriter) reset() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.written = 0
+	w.start = time.Now()
+	w.measuredWindows++
+}
+
+func (w *TransferSpeedMonitoringWriter) checkSpeed(min float64) bool {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	d := time.Now().Sub(w.start)
```

Comment:
(thinking out loud) I wonder if we should protect against "d.Seconds()" being "0" to avoid division by zero? I know it's highly unlikely but it would crash snapd if this would ever happen

Reply:
I suppose just return ok = true in that case

Reply:
Added just in case; this won't happen right now because we only call checkSpeed on time.After(w.measureTimeWindow) in Monitor().